src/management-system-v2/lib/user-preferences.ts

```diff
@@ -54,6 +42,119 @@ const defaultPreferences = {
   'environments-page-side-panel': { open: false, width: 300 },
 }; /* as const */ /* Does not work for strings */
 
+const partialUpdate = (
+  defaultPreferences: PreferencesType,
+  currentPreferences: PreferencesType,
+): PreferencesType => {
+  const updateArray = (defaultArray: any[], currentArray: any[]): any[] => {
+    /* In case default array is empty */
+    if (defaultArray.length === 0) return currentArray;
+
+    /* Shape-Switch */
+    let shapeIsDifferent = false;
+
+    /* just take first element of array, since only the inner shape and not the values are interesting */
+    const defaultValueSample =
+      defaultArray[0]; /* NOTE: Assumes that every default array has at least one entry! (Check above) */
+
+    const updatedArray = currentArray.map((currentValue) => {
+      /* For speed up */
+      if (shapeIsDifferent) return currentValue;
+
+      /* Default elment is an array */
+      if (Array.isArray(defaultValueSample)) {
+        /* Currently saved is not an array */
+        if (!Array.isArray(currentValue)) {
+          shapeIsDifferent = true;
+          return defaultValueSample;
+          /* Currently saved is an array */
+        } else {
+          return updateArray(defaultValueSample, currentValue);
+        }
+        /* Default elment is an an object */
+      } else if (typeof defaultValueSample === 'object' && defaultValueSample !== null) {
+        /* Currently saved is not an object */
+        if (
+          typeof currentValue !== 'object' ||
+          currentValue === null ||
+          Array.isArray(currentValue)
+        ) {
+          shapeIsDifferent = true;
+          return defaultValueSample;
+          /* Currently saved is an object */
+        } else {
+          /* Shape is conform with default shape (first element) */
+          try {
+            const result = updateObject(defaultValueSample, currentValue);
+            return result;
+            /* Shape is not conform with default shape */
+          } catch (error) {
+            shapeIsDifferent = true;
+            return defaultValueSample;
+          }
+        }
+        /* Everything else */
+      } else {
+        return currentValue;
+      }
+    });
+
+    return shapeIsDifferent ? defaultArray : updatedArray;
+  };
+
+  const updateObject = (
+    defaultObject: Record<string, any>,
+    currentObject: Record<string, any>,
+    root: boolean = false,
+  ): Record<string, any> => {
+    let updatedObject: Record<string, any> = {};
+    /* Remove entries not present in default */
+    for (const key in defaultObject) {
+      updatedObject[key] = currentObject.hasOwnProperty(key)
+        ? currentObject[key]
+        : defaultObject[key];
+    }
```

Comment:
Could this be done below after the array and object checks, so we don't iterate over it twice?

Reply:
Yes, you are correct. half of the loops logic was part of the second loop already anywayse. Overlooked the other part. It's just one loop now.